### ISSUE TITLE
Substream server unable to return album art for app ultrasonic

### ISSUE THREAD

**Issue by Cknight70** [owncloud/music#896]
Hello, I am unable to download album art with the app ultrasonic as it will return error code 500 when requesting. 
They have advised me this is an upstream issue, when I opened an issue on their project here https://github.com/ultrasonic/ultrasonic/issues/579#issuecomment-917681420 

I am using music version 1.3.3 with nextcloud version 22.1.1.

Happy to provide any additional details.

**Comment by paulijar** [comment]
Thanks for the report. What do you see in nextcloud.log after the unsuccessful attempt to download the images? Are the album covers shown without problem on the Music web UI?

**Comment by Cknight70** [comment]
Album covers are shown in the web interface and successfully download in at least Strawberry and Audinaught.

Here is one of the many error logs related to Ultrasonic requesting album art:
```
{"reqId":"4rVa1L7G680gD3zXJVRq","level":3,"time":"2021-09-12T07:49:03+00:00","remoteAddr":"172.28.0.1","user":"--","app":"index","method":"GET","url":"/apps/music/subsonic/rest/getCoverArt.view?id=album-1939&size=2076&u=Myrealusername&c=Ultrasonic&f=json&v=1.13.0&p=enc:realapikey","message":"OCA\\Music\\Controller\\SubsonicController::getCoverArt(): Argument #2 ($size) must be of type ?int, string given, called in /var/www/html/custom_apps/music/lib/Controller/SubsonicController.php on line 179","userAgent":"okhttp/3.12.13","version":"22.1.1.2","exception":{"Exception":"Exception","Message":"OCA\\Music\\Controller\\SubsonicController::getCoverArt(): Argument #2 ($size) must be of type ?int, string given, called in /var/www/html/custom_apps/music/lib/Controller/SubsonicController.php on line 179","Code":0,"Trace":[{"file":"/var/www/html/lib/private/AppFramework/App.php","line":156,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\Music\\Controller\\SubsonicController"},"handleRequest"]},{"file":"/var/www/html/lib/private/Route/Router.php","line":301,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["SubsonicController","handleRequest",{"__class__":"OC\\AppFramework\\DependencyInjection\\DIContainer"},{"method":"getCoverArt.view","_route":"music.subsonic.handleRequest"}]},{"file":"/var/www/html/lib/base.php","line":1000,"function":"match","class":"OC\\Route\\Router","type":"->","args":["/apps/music/subsonic/rest/getCoverArt.view"]},{"file":"/var/www/html/index.php","line":36,"function":"handleRequest","class":"OC","type":"::","args":[]}],"File":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","Line":158,"Previous":{"Exception":"TypeError","Message":"OCA\\Music\\Controller\\SubsonicController::getCoverArt(): Argument #2 ($size) must be of type ?int, string given, called in /var/www/html/custom_apps/music/lib/Controller/SubsonicController.php on line 179","Code":0,"Trace":[{"file":"/var/www/html/custom_apps/music/lib/Controller/SubsonicController.php","line":179,"function":"getCoverArt","class":"OCA\\Music\\Controller\\SubsonicController","type":"->","args":["album-1939","2076"]},{"file":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","line":217,"function":"handleRequest","class":"OCA\\Music\\Controller\\SubsonicController","type":"->","args":["getCoverArt"]},{"file":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","line":126,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\Music\\Controller\\SubsonicController"},"handleRequest"]},{"file":"/var/www/html/lib/private/AppFramework/App.php","line":156,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\Music\\Controller\\SubsonicController"},"handleRequest"]},{"file":"/var/www/html/lib/private/Route/Router.php","line":301,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["SubsonicController","handleRequest",{"__class__":"OC\\AppFramework\\DependencyInjection\\DIContainer"},{"method":"getCoverArt.view","_route":"music.subsonic.handleRequest"}]},{"file":"/var/www/html/lib/base.php","line":1000,"function":"match","class":"OC\\Route\\Router","type":"->","args":["/apps/music/subsonic/rest/getCoverArt.view"]},{"file":"/var/www/html/index.php","line":36,"function":"handleRequest","class":"OC","type":"::","args":[]}],"File":"/var/www/html/custom_apps/music/lib/Controller/SubsonicController.php","Line":400},"CustomMessage":"--"},"id":"613e5dab015a3"}
```

**Comment by paulijar** [comment]
Which version of PHP are you using? I currently have no test environment with PHP versions newer than 7.4, but this version is emitting deprecation warnings when using the Subsonic API. So the code might not be compatible with PHP 8.x.

**Comment by Cknight70** [comment]
I am using the official docker image. 
```
# docker exec -it 894a142c77d4 bash
# php -version
PHP 8.0.10 (cli) (built: Aug 26 2021 21:50:22) ( NTS )
Copyright (c) The PHP Group
Zend Engine v4.0.10, Copyright (c) Zend Technologies
    with Zend OPcache v8.0.10, Copyright (c), by Zend Technologies
```

**Comment by paulijar** [comment]
@Cknight70 I fixed the deprecated use of a system API. Could you test if this solves the issue for you? You should download https://raw.githubusercontent.com/owncloud/music/master/lib/AppFramework/Utility/RequestParameterExtractor.php and replace the corresponding file under you Music app installation with it.

**Comment by Cknight70** [comment]
I tested this and I can now see the album art of the majority of my albums with ultrasonic! 

I don't know for sure if this is considered to be the same bug and I'd be happy to open a new issue, but some of my flac albums do not appear to have album art in the web app or in ultra sonic which I know do have album covers.

I saw this album related error in my logs as I re-scanned my music today

```
"reqId":"fgXHkwxWrXhZlqB2K3A7","level":3,"time":"2021-09-13T17:17:00+00:00","remoteAddr":"172.28.0.1","user":"myrealusername","app":"PHP","method":"POST","url":"/apps/music/api/scan","message":"usort(): Returning bool from comparison function is deprecated, return an integer less than, equal to, or greater than zero at /var/www/html/custom_apps/music/lib/Db/AlbumMapper.php#343","userAgent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.159 Safari/537.36","version":"22.1.1.2","exception":{"Exception":"Error","Message":"usort(): Returning bool from comparison function is deprecated, return an integer less than, equal to, or greater than zero at /var/www/html/custom_apps/music/lib/Db/AlbumMapper.php#343","Code":0,"Trace":[{"function":"onError","class":"OC\\Log\\ErrorHandler","type":"::","args":[8192,"usort(): Returning bool from comparison function is deprecated, return an integer less than, equal to, or greater than zero","/var/www/html/custom_apps/music/lib/Db/AlbumMapper.php",343]},{"file":"/var/www/html/custom_apps/music/lib/Db/AlbumMapper.php","line":343,"function":"usort","args":[[{"fileid":"403806","name":"AlbumArtSmall.jpg"},{"fileid":"403808","name":"Folder.jpg"},{"fileid":"403807","name":"cover.jpg"}],{"__class__":"Closure"}]},{"file":"/var/www/html/custom_apps/music/lib/BusinessLayer/AlbumBusinessLayer.php","line":282,"function":"findAlbumCover","class":"OCA\\Music\\Db\\AlbumMapper","type":"->","args":[15297,403618]},{"file":"/var/www/html/custom_apps/music/lib/Utility/Scanner.php","line":627,"function":"findCovers","class":"OCA\\Music\\BusinessLayer\\AlbumBusinessLayer","type":"->","args":["myrealusername"]},{"file":"/var/www/html/custom_apps/music/lib/Controller/ApiController.php","line":382,"function":"findAlbumCovers","class":"OCA\\Music\\Utility\\Scanner","type":"->","args":["myrealusername"]},{"file":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","line":217,"function":"scan","class":"OCA\\Music\\Controller\\ApiController","type":"->","args":["405580,405582,405584,405586",true]},{"file":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","line":126,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\Music\\Controller\\ApiController"},"scan"]},{"file":"/var/www/html/lib/private/AppFramework/App.php","line":156,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\Music\\Controller\\ApiController"},"scan"]},{"file":"/var/www/html/lib/private/Route/Router.php","line":301,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["ApiController","scan",{"__class__":"OC\\AppFramework\\DependencyInjection\\DIContainer"},{"_route":"music.api.scan"}]},{"file":"/var/www/html/lib/base.php","line":1000,"function":"match","class":"OC\\Route\\Router","type":"->","args":["/apps/music/api/scan"]},{"file":"/var/www/html/index.php","line":36,"function":"handleRequest","class":"OC","type":"::","args":[]}],"File":"/var/www/html/lib/private/Log/ErrorHandler.php","Line":92,"CustomMessage":"--"},"id":"613f8d26a8105"}
```

```
Error: usort(): Returning bool from comparison function is deprecated, return an integer less than, equal to, or greater than zero at /var/www/html/custom_apps/music/lib/Db/AlbumMapper.php#343
<<closure>>

OC\Log\ErrorHandler::onError(8192, "usort(): Re ... o", "/var/www/ht ... p", 343)

/var/www/html/custom_apps/music/lib/Db/AlbumMapper.php - line 343:

usort([ { fileid: ... }], Closure {})

/var/www/html/custom_apps/music/lib/BusinessLayer/AlbumBusinessLayer.php - line 282:

OCA\Music\Db\AlbumMapper->findAlbumCover(16288, 404969)

/var/www/html/custom_apps/music/lib/Utility/Scanner.php - line 627:

OCA\Music\BusinessLayer\AlbumBusinessLayer->findCovers("myrealusername")

/var/www/html/custom_apps/music/lib/Controller/ApiController.php - line 382:

OCA\Music\Utility\Scanner->findAlbumCovers("myrealusername")

/var/www/html/lib/private/AppFramework/Http/Dispatcher.php - line 217:

OCA\Music\Controller\ApiController->scan("405580,405582,405584,405586", true)

/var/www/html/lib/private/AppFramework/Http/Dispatcher.php - line 126:

OC\AppFramework\Http\Dispatcher->executeController(OCA\Music\Co ... {}, "scan")

/var/www/html/lib/private/AppFramework/App.php - line 156:

OC\AppFramework\Http\Dispatcher->dispatch(OCA\Music\Co ... {}, "scan")

/var/www/html/lib/private/Route/Router.php - line 301:

OC\AppFramework\App::main("ApiController", "scan", OC\AppFramew ... {}, { _route: "music.api.scan"})

/var/www/html/lib/base.php - line 1000:

OC\Route\Router->match("/apps/music/api/scan")

/var/www/html/index.php - line 36:

OC::handleRequest()
```

**Comment by paulijar** [comment]
Your second problem seems to be totally distinct from the first one, so a new issue should be opened. My best guess is that this time the deprecation error in the log is not actually related to the missing album art, but it's still something to be fixed. Are those flac files supposed to have embedded album art or a separate image file residing in the same folder with the flac file?

**Comment by Cknight70** [comment]
They are embedded into the files. I have another flac album with a cover.jpg file which does show the album art of. I will make a new issue for this, but thank you for your work!

**Comment by frake65** [comment]
Can confirm this patch also helps for my installation (nextcloud 22 on PHP8), which had issues on loading all the cover art instantly. It now loads most of the cover art embedded within the mp3 files.
Thank you!

**Comment by paulijar** [comment]
The fix for this issue is now released in Music v1.4.0.
